Blog post about first steps on SCM integration (#294)
openSUSE/obs-landing: _posts/development/2021-05-31-scm-integration.md

We have started to integrate GitHub and GitLab for a scenario that seems to be the most expected one among you:
running an OBS build in your pull/merge request continuous integration cycle.
So if you follow the steps below, every new pull/merge request will trigger a package branch in OBS and
report back the status of this test build.

### Workflow Integration Steps

#### 1. Set the Configuration File

First of all, you have to configure the integration between the repository in the SCM system (where your package sources live)
and the project and package inside OBS (where the builds are going to run).

To do so, in the root directory of your GitHub/GitLab repository, you have to create a directory `.obs` which contains the
configuration file `workflows.yml`.

For example, for a repository called `my-game` in GitHub/GitLab, you might have an OBS project `home:Iggy` which has a package `ctris`.
And the configuration file should look like this:

```
workflow:
  steps:
    - branch_package:
        source_project: home:Iggy
        source_package: ctris
```

This indicates that, on every new pull/merge request in the repository's side, OBS will branch the project/package given by

> `source_project` and `source_package` fields and will try to build them.

Reviewer: Shouldn't it be enough to register the token for the package? Why do we need an additional file with the same information?

#### 2. Create Github/GitLab Personal Access Token

At some point, OBS will report the result of the workflow steps back to GitHub/GitLab.
Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and
[GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation
to learn how to create a personal access token.

In the next step, you will learn how to pass the so-called `scm_token` to OBS, this token represents your user account on the SCM system's side.

#### 3. Create a Workflow Token in OBS

As you want to trigger some actions in OBS as a response to the events hapenning GitHub/GitLab,
you also need an OBS's access token. GitHub/GitLab will use it to trigger the actions on your behalf.

The token represents your user account and the kind of action/operation you want to perform, which is `workflow` in this case.

You can create a token in OBS using `osc`:

```
osc token --create --operation workflow scm_token=ssx_j2du52Ozs8oZAtxq3K0bdy9TEp8iUy4JilTs

> ```

Reviewer: it would be good to have the example for a specific package here. The global tokens (permitting it in any package where the user has maintainer rights) should not be used by default. 

It would mean to give write access to the SCM hoster for any package, maybe osc should not even allow this by default...

Reviewer 2: The workflow tokens are never bound to a package. They are just bound to a users and what they can and can't do.

Reviewer 3: > It would mean to give write access to the SCM hoster for any package, maybe osc should not even allow this by default...

This is why people shouldn't leak their SCM token. Same for the OBS token, we also warn users about this in this blog post.

---
layout: post
title: Heading for a GitHub and GitLab Workflow Integration in OBS
category: development
---

Most of the code implemented nowadays is managed collaboratively
through source code management (SCM) systems like GitHub, GitLab or Pagure.
An increasing number of people want to handle their package sources with those systems too

> and while there is an integration between SCMs and OBS, it could definitely be improved.

Author: As `An increasing number of you want to...` sounds weird to me I've decided to write `Every day more of you want to... `

> and while there is an integration between SCMs and OBS, it could definitely be improved.

Reviewer 3: I'm fine with that change. I don't have a better suggestion.

In today's post, we are sharing how we began to move to meet that need.

## SCM to Collaborate, OBS to Build

We have started to integrate GitHub and GitLab for a scenario that seems to be the
most expected one among OBS users: managing the package sources in an SCM system

> and use OBS to build the source code and report back the results to GitHub or GitLab.

Author: Done

With the appropriate configuration, every new pull/merge request in the SCM system
will trigger a package branch in OBS. Then, as soon as the branched package's building finishes,
OBS will report back the build status to the pull/merge request.

> We propose the following steps to achieve this goal. Give them a try!

Author: Done

### 1. Set a Configuration File

In the root directory of your GitHub/GitLab repository, create a directory `.obs` which contains the configuration file `workflows.yml`.

The content of the file should look like this:

```
workflow1:
  steps:
    - branch_package:
        source_project: home:Iggy
        source_package: ctris
```

> For now, OBS only accepts one workflow with a single step: `branch_package`.

Author: This whole section has been merged with the previous one and the whole text has changed.

> Adapt the names of the source project and package to your needs.

Author: This whole section has been merged with the previous one and the whole text has changed.

> In the example above, we went with the workflow name `workflow1`. You can be more creative!

Author: This whole section has been merged with the previous one and the whole text has changed.

> The step `branch_package` indicates to OBS that you want to branch the project/package given by `source_project` and `source_package` fields.

Author: This whole section has been merged with the previous one and the whole text has changed.

The resulting branched project/package will be !!! TODO - Document this !!!

### 2. Create Project and Package in OBS

> Apart from the repository inside an SCM system, you have to create an analogous project/package inside OBS.

Author: Rephrased.

The `_service` stuff is still a TODO.

  <figcaption>GitHub Webhook Form</figcaption>
</figure>

You have to fill in the form with
- *Payload URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 4).
- *Content type*: please choose `application/json`.
- *Secret*: the OBS's token string obtained in step 4.
- *Enable SSL verification*.
- *Let me select individual events*. We just need you to select `Pull requests`.

As soon as you add the webhook, OBS will start receiving requests for every pull request happening in GitHub.

#### Set a GitLab Webhook

In GitLab, you can go to *Settings -> Webhooks*.

<figure>
  <img src="/images/posts/sprint_95_gitlab_webhook_form.png" alt="GitLab Webhook Form" />
  <figcaption>GitLab Webhook Form</figcaption>
</figure>

You have to fill in the form with
- *URL*: the OBS endpoint to trigger the workflow operation (https://build.opensuse.org/trigger/workflow?id=0). Ensure you replace the `id` with the one obtained in the token creation (step 4).
- *Secret Token*: the OBS's token string obtained in step 4.

> - *Trigger*. We just need you to select `Merge Requests`.

Reviewer 2: The screenshot doesn't reflect this.

Author: Screenshot updated, now only "Merge Requests" is selected.

---
layout: post
title: Heading for a GitHub and GitLab Workflow Integration in OBS
category: development
---

Most of the code implemented nowadays is managed collaboratively
through source code management (SCM) systems like GitHub, GitLab or Pagure.
An increasing number of you want to handle their package sources with those systems too.
And while there is an integration between SCMs and OBS, it could definitely be extended and improved.

> In today's post, we are sharing how we began to move to meet that need.

Reviewer 4: Hmm, I think we are more showing what is possible and how to use it

> In today's post, we are sharing how we began to move to meet that need.

Reviewer 4: ```suggestion
In today's post, we are sharing what's possible in these regards from now on and how to use it.```

report back the status of this test build.

### Workflow Integration Steps

#### 1. Set the Configuration File and Service

First of all, you have to configure the integration between the repository in the SCM system (where your package sources live)
and the project and package inside OBS (where the builds are going to run).

To do so, in the root directory of your GitHub/GitLab repository, you have to create a directory `.obs` which contains the
configuration file `workflows.yml`.

For example, for a repository called `my-game` in GitHub/GitLab, you might have an OBS project `home:Iggy` which has a package `ctris`.
And the configuration file should look like this:

```
workflow:
  steps:
    - branch_package:
        source_project: home:Iggy
        source_package: ctris
```

This indicates that, on every new pull/merge request in the repository's side, OBS will branch the project/package given by
`source_project` and `source_package` fields and will try to build them.

For all of this to be possible, the OBS workflows depend on the [obs-service-tar_scm](https://github.com/openSUSE/obs-service-tar_scm) service which is configured in the `_service` file of the source package.

> This service has to be configutred to point to your GitHub/GitLab repository.

Reviewer 4: ```suggestion
This service has to be configured to point to your GitHub/GitLab repository.
```
```suggestion
This service has to be configutred to point to your GitHub/GitLab repository.
```

!! TODO - Finish example of a _service file, taken from https://en.opensuse.org/openSUSE:Build_Service_Concept_SourceService#Example_2:_GIT_integration !!!

```
<service name="tar_scm">
  <param name="scm">git</param>
  <param name="url">the URL of your Git repository (something like "git://github_or_gitlab.com/your_repository.git")</param>
  <param name="subdir">the subdirectory of your repository you want to create an archive from</param>
  <param name="filename">the name of the file you want to create, without version</param>
  <param name="versionprefix">the first part of the version string (e.g. "0.4.git")</param>
  <param name="revision">optional: the commit hash, tag or branch you want to get</param>
</service>
```

> Without a service, we cannot pull the source changes in OBS, thus rendering OBS workflows pretty much useless.

Reviewer 4: @dmarcoux I wouldn't go into much details regarding the `_branch_request` file, but just thinking that it might be worth to mention that OBS will take care of the rest (like pointing to the right revision/commit once the workflow got triggered)

category: development
---

Most of the code you implement nowadays is managed collaboratively through source code management (SCM) systems
like GitHub, GitLab or Pagure. And an increasing number of you also want to handle their package sources with those systems.
While there are various ways to integrate OBS into your SCM, they can definitely be extended and improved.
In today's post, we are sharing the first beta for a deeper integration between SCMs and OBS. 

We have started to implement a scenario that seems to be the most expected one among you:
running an OBS build in your pull/merge request continuous integration cycle. SCM to collaborate, OBS to build.

We are following the process you probably already know from setting up other continuous integration tools:

1. You allow OBS and Github/Gitlab to talk to each other via tokens
1. Github/Gitlab tells OBS that a new pull/merge request was opened via webhooks
1. You configure the workflow that should run on OBS in a YAML file in your Github/Gitlab repository
1. OBS executes the workflow you configured
1. OBS reports the status (pending, success or failure) of the workflow to the pull/merge request that was opened

> There are some prerequisites we assume you already have already. First, if you haven't done it already, you have to join [the beta program](https://openbuildservice.org/2018/10/04/the-beta-program/), so you can try this new feature. And surprise surprise, you will need a repository on Github/Gitlab and a package on build.opensuse.org you want to integrate! 🤓

Author: ```suggestion
There are some prerequisites we assume you already have. First, if you haven't done it yet, you have to join [the beta program](https://openbuildservice.org/2018/10/04/the-beta-program/), so you can try this new feature. And surprise surprise, you will need a repository on Github/Gitlab and a package on build.opensuse.org you want to integrate! 🤓
```

Let's set things up.

## Authentication Setup

> First of all we have to make sure that OBS and your SCM service are *allowed* to talk to each other. Tokens are the

Author: ```suggestion
First of all, we have to make sure that OBS and your SCM service are *allowed* to talk to each other. Tokens are the
```

way to make this happen.

### 1. Create a Github/GitLab Personal Access Token

You have to give OBS permissions to report the build status to the pull/merge request to the SCM on your behalf.

> For this you have to create a personal access token. Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and [GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation to learn how.

Author: ```suggestion
For this, you have to create a personal access token. Check [GitHub's](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) and [GitLab's](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#creating-a-personal-access-token) documentation to learn how.
```

> Make sure the access token you generate has the *at least* those scopes assigned:

Author: ```suggestion
Make sure the access token you generate has *at least* those scopes assigned:
```

- **Github**: `repo`
- **Gitlab**: `api`

Note down this access token, you are going to use it in the next step. For the sake of clarity, let's pretend your Github/Gitlab access token is: `abcdefg`

### 2. Create a Workflow Token in OBS

You also have to give your SCM permissions to trigger a workflow inside OBS on your behalf.
That's why you need to create an OBS access token.

You can create one using `osc`:

```
$ osc token --create --operation workflow scm_token=abcdefg
Create a new token
<status code="ok">
  <summary>Ok</summary>
  <data name="token">uvwxyz</data>
  <data name="id">12345</data>
</status>
```

> Of course you should provide your real SCM personal access token here, created in the previous step.

Author: ```suggestion
Of course, you should provide your real SCM personal access token here, created in the previous step.
```

By specifying the operation `workflow` you limit the scope of this token to this kind of operation.

> Note down this access token and it's ID, you are going to use them in the next step. For the sake of

Author: ```suggestion
Note down this access token and its ID, you are going to use them in the next step. For the sake of
```

> clarity, let's pretend your OBS access token is: `uvwxyz` and it's numerical ID is `12345`.

Author: ```suggestion
clarity, let's pretend your OBS access token is: `uvwxyz` and its numerical ID is `12345`.
```

**WARNING:** Tokens are like the keys to (parts of) your house. You have to keep your token secret
to prevent someone else from triggering operations in your name!

You can always revoke your token and create a new one if you suspect it was leaked by you somehow.

```
osc token # list all your tokens
osc token --delete <token_id> # remove the token with the given id
```

By the way, there are also various other types of tokens in the OBS that can help you to authenticate for
operations in OBS. Check out [the manual](https://openbuildservice.org/help/manuals/obs-user-guide/cha.obs.authorization.token.html#id-1.5.10.16.4) if you're interested.

> ## Setup a Webhook

Author: ```suggestion
## Set up a Webhook
```

### On GitHub

In GitHub, you can go to *Settings -> Webhooks*.

<figure>
  <img src="/images/posts/sprint_95_github_webhook_form.png" alt="GitHub Webhook Form" />
  <figcaption>GitHub Webhook Form</figcaption>
</figure>

You have to fill in the form with
- **Payload URL**: `https://build.opensuse.org/trigger/workflow?id=12345`.
  <br /> Replace `12345` with the OBS token numerical ID obtained above.
- **Content type**: `application/json`
- **Secret**: `uvwxyz`.
  <br /> Replace `uvwxyz` with the OBS token secret string obtained above.
- **Enable SSL verification**.
- **Let me select individual events**: `Pull requests`

### On GitLab

In GitLab, you can go to *Settings -> Webhooks*.

<figure>
  <img src="/images/posts/sprint_95_gitlab_webhook_form.png" alt="GitLab Webhook Form" />
  <figcaption>GitLab Webhook Form</figcaption>
</figure>

You have to fill in the form with
- **URL**: https://build.opensuse.org/trigger/workflow?id=12345
  <br /> Replace `12345` with the OBS token numerical ID obtained above.
- **Secret Token**: `uvwxyz`
  <br /> Replace `uvwxyz` with the OBS token secret string obtained above.
- **Trigger**: `Merge Requests`

> ## Setup the OBS Workflow

Author: ```suggestion
## Set up the OBS Workflow
```

Now that the two services talk to each other, we need to configure what should happen on OBS when a new
pull/merge request is opened. To do so, in the root directory of your GitHub/GitLab repository, create a directory

> `.obs` which contains the a file called `workflows.yml`.

Author: ```suggestion
`.obs` which contains a file called `workflows.yml`.
```

The content of the file should be:

```
workflow:
  steps:
    - branch_package:
        source_project: home:hennevogel:myfirstproject
        source_package: ctris
```

> - **source_project**: Configure the name of the project that containts the package you want to test build

Author: ```suggestion
- **source_project**: Configure the name of the project that contains the package you want to test build
```

> - **source_package**: Confoigure the name of the package you want to test build

Author: ```suggestion
- **source_package**: Configure the name of the package you want to test build
```

the project `home:hennevogel:myfirstproject` and build it. The result OBS will report back to Github or Gitlab.

<figure>
  <img src="/images/posts/sprint_95_checks_github.png" alt="Checks GitHub" />
  <figcaption>Checks GitHub</figcaption>
</figure>

<figure>
  <img src="/images/posts/sprint_95_checks_gitlab.png" alt="Checks GitLab" />
  <figcaption>Checks GitLab</figcaption>
</figure>

## Use the SCM sources in your OBS package

We have come very far! The services talk to each other, OBS knows what to do and where to report to. But that's not
all right? You probably also want to make the package, you are testing, use the sources from the pull/merge request.

For this, you can make use of the existing [obs-service-tar_scm](https://github.com/openSUSE/obs-service-tar_scm#user-documentation) service. obs-service-tar_scm will automatically use the sources of the pull/merge request that triggered it: Magic 🪄

## Current Limitations

As you might notice, in this first beta, we are just triggering operations right after a new pull/merge request is
created. What happens if a new commit is added to the pull/merge request or if someone force pushes some changes?
We don't handle this yet.

> Depending on the package you branch, it might build for more than one repository or architecture. Currently

Author: ```suggestion
Depending on the package your branch, it might build for more than one repository or architecture. Currently
```

Reviewer 3: The version with `you` is correct.

---
layout: post
title: Continuous Integration with OBS and GitHub/GitLab
category: development
---

Most of the code you implement nowadays is managed collaboratively through source code management (SCM) systems

> like GitHub, GitLab or Pagure. And an increasing number of you also want to handle their package sources with those systems.

Reviewer 5: ```suggestion
like GitHub, GitLab or Pagure. An increasing number of you also want to handle their package sources with those systems.
```

Reviewer 5: And maybe join that sentence with the previous one

Reviewer 3: I find it good as it is.

Reviewer 5: Okay!

  <figcaption>Checks GitHub</figcaption>
</figure>

<figure>
  <img src="/images/posts/sprint_95_checks_gitlab.png" alt="Checks GitLab" />
  <figcaption>Checks GitLab</figcaption>
</figure>

## Use the SCM Sources in Your OBS package

We have come very far! The services talk to each other, OBS knows what to do and where to report to.
But that's not all, right? You probably also want to make the package, you are testing, use the sources from the pull/merge request.

For this, you can make use of the existing [obs-service-tar_scm](https://github.com/openSUSE/obs-service-tar_scm#user-documentation) service.
obs-service-tar_scm will automatically use the sources of the pull/merge request that triggered it: Magic 🪄

## Current Limitations

As you might notice, in this first beta, we are just triggering operations right after a new pull/merge request is
created. What happens if a new commit is added to the pull/merge request or if someone force pushes some changes?
We don't handle this yet.

Depending on the package you branch, it might build for more than one repository or architecture. Currently
we will report the *last* build result that occurs.

> 
> [//]: # FIXME: is that so?
> We also don't report back the result of a manually triggered rebuild in OBS.

Reviewer 5: ```suggestion
```
